File: android/src/main/java/com/reactnativeescposprinter/ThePrinter.java
package com.escposprinter;

import com.epson.epos2.Epos2Exception;
import com.epson.epos2.printer.Printer;
import com.epson.epos2.printer.PrinterSettingListener;
import com.epson.epos2.printer.PrinterStatusInfo;
import com.epson.epos2.printer.ReceiveListener;
import com.epson.epos2.printer.StatusChangeListener;
import com.epson.epos2.Epos2CallbackCode;

import com.escposprinter.PrinterDelegate;
import com.escposprinter.EposStringHelper;
import com.escposprinter.PrinterCallback;
import com.escposprinter.ImageManager;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.Arguments;

import android.content.Context;
import android.graphics.Bitmap;
import android.util.Base64;

public class ThePrinter implements PrinterSettingListener, ReceiveListener {

    private Printer epos2Printer_ = null; // Printer
    private volatile String printerTarget_ = null; // the printer target
    private boolean isConnected_ = false; // cache if printer is connected
    private boolean didBeginTransaction_ = false; // did start Status Monitor
    private static final int DISCONNECT_INTERVAL = 500;//millseconds

    private PrinterCallback printCallback_ = null; // callback
    private PrinterCallback getPrinterSettingCallback_ = null; // callback

    /**
     * Returns ThePrinter
     * Use to create ThePrinter object
     */
    public ThePrinter() {

    }

    /**
     * Dealloc ThePrinter
     */
    public void finalize() {
        printerTarget_ = null;
    }

    /**
     Returns void
     function initWith Will create a new printer with settings given
     @param printerTarget the target for the printer  -- deviceInfo.target
     @param series the printer series -- EPOS2_TM_M30II
     @param lang the printer language -- EPOS2_MODEL_ANK
     @param callback delegate callbacks
     returns void
     */
    synchronized public void setupWith(final String printerTarget, final int series, final int lang, Context context) throws Epos2Exception {
        printerTarget_ = printerTarget;
        epos2Printer_ = new Printer(series, lang, context);
        epos2Printer_.setReceiveEventListener(this);
    }


    /**
     Returns String
     Function getPrinterTarget retrieves the printer target set from initWith.
     @return String containing printer target
     */
    public String getPrinterTarget() {
        return printerTarget_;
    }

    /**
     Returns BOOL
     Function isConnected uses getStatus to understand if it is connected or not.
     @return bool -- returns true if connected
     */
     synchronized public boolean isConnected()
    {
        // check to see if we are actually connected

        if (epos2Printer_ == null) return false;

        boolean isConnected = true;
        PrinterStatusInfo info = epos2Printer_.getStatus();
        if (info.getConnection() == Printer.TRUE) {
            isConnected = true;
        } else {
            isConnected = false;
        }

        return isConnected;
    }

    /**
     throws Epos2Exception if there is an error
     Function connect tries to connect selected printer
     @param timeout the amount of time before giving up -- EPOS2_PARAM_DEFAULT
     */
    public void connect(final int timeout) throws Epos2Exception {

        synchronized (this) {
            if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);
            if (printerTarget_ == null || printerTarget_.isEmpty())
                throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

            if (isConnected_) {
                return;
            }

            isConnected_ = false;

            try {
                epos2Printer_.connect(printerTarget_, timeout);
                isConnected_ = true;
            } catch (Epos2Exception e) {
                e.printStackTrace();
                isConnected_ = false;
                throw e;
            }
        }
    }

    /**
     throws Epos2Exception if there is an error
     Function disconnect tries to disconnect selected printer
     */
    synchronized public void disconnect() throws Epos2Exception {

        if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

        if (!isConnected_) return;

        if (didBeginTransaction_) {
            try {
                endTransaction();
            } catch (Epos2Exception e) {
                e.printStackTrace();
            }
        }

        int count = 20;
        while (true) {
            try {
                epos2Printer_.disconnect();
                break;
            } catch (final Exception e) {
                if (e instanceof Epos2Exception) {
                    //Note: If printer is processing such as printing and so on, the disconnect API returns ERR_PROCESSING.
                    if (((Epos2Exception) e).getErrorStatus() == Epos2Exception.ERR_PROCESSING) {
                        count--;
                        try {
                           Thread.sleep(DISCONNECT_INTERVAL);
                        } catch (Exception ignored) {
                        }
                    } else {
                      if(((Epos2Exception) e).getErrorStatus() != Epos2Exception.ERR_ILLEGAL) {
                         throw e;
                      }
                    }
                }
            }
            if (count == 0) {
                throw new Epos2Exception(Epos2Exception.ERR_DISCONNECT);
            }
        }

        epos2Printer_.clearCommandBuffer();
        isConnected_ = false;
        didBeginTransaction_ = false;

    }

    synchronized public void clearCommandBuffer() throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

      epos2Printer_.clearCommandBuffer();
    }

    synchronized public void sendData(int timeout, PrinterCallback handler) throws Epos2Exception {
        if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

        beginTransaction();
        try {
           epos2Printer_.sendData(timeout);
           printCallback_ = handler;
        } catch (Epos2Exception e) {
           endTransaction();
           throw e;
        }

    }

    /**
     throws Epos2Exception if there is an error
     Function beginTransaction see ePOS SDK
     */
    synchronized public void beginTransaction() throws Epos2Exception {
        if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

        if (didBeginTransaction_) return;

        try {
            epos2Printer_.beginTransaction();
            didBeginTransaction_  = true;
        } catch (Epos2Exception e) {
            didBeginTransaction_ = false;
            e.printStackTrace();
            throw e;
        }
    }

    /**
     throws Epos2Exception if there is an error
     Function endTransaction see ePOS SDK
     */
    synchronized public void endTransaction() throws Epos2Exception {

        if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

        if (!didBeginTransaction_) return;

        try {
            epos2Printer_.endTransaction();
            didBeginTransaction_  = false;
        } catch (Epos2Exception e) {
            e.printStackTrace();
            throw e;
        }
    }

    synchronized public void addText(String data) throws Epos2Exception {
        if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

        epos2Printer_.addText(data);
    }


    synchronized public void addTextLang(int lang) throws Epos2Exception {
        if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

        epos2Printer_.addTextLang(lang);
    }

    synchronized public void addFeedLine(int line) throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

      epos2Printer_.addFeedLine(line);
    }

    synchronized public void addLineSpace(int linespc) throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

      epos2Printer_.addLineSpace(linespc);
    }

    synchronized public void addCut(int type) throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

      epos2Printer_.addCut(type);
    }

    synchronized public void addCommand(String base64string) throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

      byte[] data = Base64.decode(base64string, Base64.DEFAULT);
      epos2Printer_.addCommand(data);
    }

    synchronized public void addPulse(int drawer, int time) throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

      epos2Printer_.addPulse(drawer, time);
    }

    synchronized public void addTextAlign(int align) throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

      epos2Printer_.addTextAlign(align);
    }

    synchronized public void addTextSize(int width, int height) throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

      epos2Printer_.addTextSize(width, height);
    }

    synchronized public void addTextSmooth(int smooth) throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

      epos2Printer_.addTextSmooth(smooth);
    }

    synchronized public void addTextStyle(int reverse, int ul, int em, int color) throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

      epos2Printer_.addTextStyle(reverse, ul, em, color);
    }

    synchronized public void addImage(ReadableMap source, Context mContext, int width, int color,
                                      int mode, int halftone, double brightness, int compress) throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

      Bitmap bitmap = null;

      try {
        bitmap = ImageManager.getBitmapFromSource(source, mContext);
      } catch (Exception e) {
        throw new Epos2Exception(Epos2Exception.ERR_MEMORY);
      }

      float aspectRatio = bitmap.getWidth() / (float) bitmap.getHeight();
      int newHeight = Math.round(width / aspectRatio);
      bitmap = Bitmap.createScaledBitmap(bitmap, width, newHeight, false);

      epos2Printer_.addImage(bitmap, 0, 0, width, newHeight, color, mode, halftone, brightness, compress);

    }

    synchronized public void addBarcode(String data, int type, int hri, int font, int width, int height) throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

      epos2Printer_.addBarcode(data, type, hri, font, width, height);
    }

    synchronized public void addSymbol(String data, int type, int level, int width, int height, int size) throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

      epos2Printer_.addSymbol(data, type, level, width, height, size);
    }

    synchronized public WritableMap getStatus() throws Epos2Exception {
      if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);
      PrinterStatusInfo status = epos2Printer_.getStatus();
      WritableMap statusMap = EposStringHelper.convertStatusInfoToWritableMap(status);
      return statusMap;
    }

  /**
     throws Epos2Exception if there is an error
     Function getPrinterSetting see ePOS SDK
     */
    synchronized public void getPrinterSetting(int timeout, int type, PrinterCallback handler) throws Epos2Exception {
        if (epos2Printer_ == null) throw new Epos2Exception(Epos2Exception.ERR_MEMORY);

        beginTransaction();
        try {
            epos2Printer_.getPrinterSetting(timeout, type, this);
            getPrinterSettingCallback_ = handler;
        } catch (Epos2Exception e) {
            endTransaction();
            throw e;
        }
    }

    /**
     returns Printer
     Function getEpos2Printer
     */
    synchronized public Printer getEpos2Printer()
    {
        return epos2Printer_;
    }

    @Override
    public void onGetPrinterSetting(int code, int type, int value) {

        new Thread(new Runnable() {
            @Override
            synchronized public void run() {
                try {
                    endTransaction();
                } catch (Exception e) {
                    e.printStackTrace();
                }
                if (getPrinterSettingCallback_ != null) {
                  if(code == Epos2CallbackCode.CODE_SUCCESS) {
                   WritableMap returnData = Arguments.createMap();
                    returnData.putInt("type", type);
                    returnData.putInt("value",value);

                    getPrinterSettingCallback_.onSuccess(returnData);
                  } else {
                    getPrinterSettingCallback_.onError(EposStringHelper.getErrorTextData(code, "code"));
                  }
                  getPrinterSettingCallback_ = null;
                }
            }
        }).start();
    }

    // region ReceiveListener
    @Override
    public void onPtrReceive(Printer printer, int code, PrinterStatusInfo status, String printJobId) {

        new Thread(new Runnable() {
            @Override
            synchronized public void run() {
                try {
                    endTransaction();
                } catch (Exception e) {
                    e.printStackTrace();
                }
                if (printCallback_ != null) {
                  if(code == Epos2CallbackCode.CODE_SUCCESS) {
                    WritableMap returnData = EposStringHelper.convertStatusInfoToWritableMap(status);
                    printCallback_.onSuccess(returnData);
                  } else {
                    printCallback_.onError(EposStringHelper.getErrorTextData(code, "code"));
                  }
                  printCallback_ = null;
                }
            }
        }).start();

    }


    @Override
    public void onSetPrinterSetting(int code) {

    }

}

File: package/ios/ThePrinter.m
#import <Foundation/Foundation.h>
#import "ThePrinter.h"
#import "EposStringHelper.h"
#import "EposStringHelper.h"
#import "ImageManager.h"


@interface ThePrinter() <Epos2PtrReceiveDelegate, Epos2PrinterSettingDelegate>
{
    void (^onPtrRecieveSuccessHandler_)(NSDictionary* data);
    void (^onPtrRecieveErrorHandler_)(NSString* data);

    void (^onGetPrinterSettingSuccessHandler_)(NSDictionary* data);
    void (^onGetPrinterSettingErrorHandler_)(NSString* data);
}
@end


@implementation ThePrinter

-(id)init {
    self = [super init];
    if (self)
    {
        printerTarget_ = nil;
        epos2Printer_ = nil;
        isConnected_ = false;
        didBeginTransaction_ = false;

        onPtrRecieveSuccessHandler_ = nil;
        onPtrRecieveErrorHandler_ = nil;

        onGetPrinterSettingSuccessHandler_ = nil;
        onGetPrinterSettingErrorHandler_ = nil;

    }
    return self;
}

- (id _Nonnull) initWith:(nonnull NSString*)printerTarget series:(int)series lang:(int)lang delegate:(id<PrinterDelegate >_Nullable)delegate
{

    @synchronized (self) {
        self = [self init];

        // store printer target
        printerTarget_ = printerTarget;

        // create printer object
        epos2Printer_ = [[Epos2Printer alloc] initWithPrinterSeries:series lang:lang];
        [epos2Printer_ setReceiveEventDelegate:self];

        _Delegate = delegate;

        // give cpu some time
        [NSThread sleepForTimeInterval:0.01];

        return self;
    }
}

- (Epos2Printer*) getEpos2Printer
{
    @synchronized (self) {
        return epos2Printer_;
    }
}

- (bool)isConnected
{
    // check to see if we are actually connected
    @synchronized (self) {
        bool isConnected = true;
        Epos2PrinterStatusInfo* info = [epos2Printer_ getStatus];
        if (info.connection) {
            isConnected = true;
        } else {
            isConnected = false;
        }

        return isConnected;
    }
}

- (nonnull NSString*) getPrinterTarget
{
    return printerTarget_;
}

#pragma mark - Epos2Printer objc API
- (int) connect:(long)timeout {
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return false;
        }

        int connectResult = EPOS2_SUCCESS;

        if (isConnected_) {
            return connectResult;
        }

        // set printer to connecting
        connectResult = [epos2Printer_ connect:printerTarget_ timeout:timeout];
        if (connectResult != EPOS2_SUCCESS)
        {
            // seet printer to idle
            isConnected_ = false;
            return connectResult;
        }

        // connection success
        isConnected_ = true;

        [NSThread sleepForTimeInterval:0.01];

        return connectResult;
    }
}

- (int) disconnect
{
    @synchronized (self) {

        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }


        if (!isConnected_) {
            return EPOS2_SUCCESS;
        }

        [self endTransaction];



        int result = 1;
        bool exit_loop = false;
        int nWaitSeconds = 10;

        // try and disconnect from printer
        while(result || exit_loop)
        {
            if(epos2Printer_ == nil)
            {
                break;
            }

            [self endTransaction];
            result = [epos2Printer_ disconnect];

            switch(result)
            {
                case EPOS2_SUCCESS:
                    exit_loop = true;
                    break;
                case EPOS2_ERR_ILLEGAL:
                    exit_loop = true;
                    break;
                case EPOS2_ERR_MEMORY:
                    exit_loop = true;
                    break;
                case EPOS2_ERR_FAILURE:
                    exit_loop = true;
                    break;
                case EPOS2_ERR_PROCESSING:
                    exit_loop = false;
                    break;
                case EPOS2_ERR_DISCONNECT:
                    exit_loop = true;
                    break;
            }

            if (exit_loop) {
                break;
            }

            [NSThread sleepForTimeInterval:1];

            if (--nWaitSeconds == 0) {
                // timeout 20aeconds.
                break;
            }
        }

        [epos2Printer_ clearCommandBuffer];
        isConnected_ = false;

        return result;
    }
}


- (int) clearCommandBuffer
{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        [epos2Printer_ clearCommandBuffer];
        return EPOS2_SUCCESS;
    }
}

- (void) sendData:(long)timeout
         successHandler: (void(^)(NSDictionary* data)) successHandler
         errorHandler: (void(^)(NSString* data)) errorHandler
{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            errorHandler([EposStringHelper convertDictionatyToJsonString:@{
                @"data": @(EPOS2_ERR_MEMORY),
                @"type": @"result"
            }]);

              return;
        };
        [self beginTransaction];
        int result =  [epos2Printer_ sendData:timeout];

        if(result == EPOS2_SUCCESS) {
            onPtrRecieveSuccessHandler_ = [successHandler copy];
            onPtrRecieveErrorHandler_ = [errorHandler copy];
        } else {
            [self endTransaction];
            errorHandler([EposStringHelper convertDictionatyToJsonString:@{
                @"data": @(result),
                @"type": @"result"
            }]);
        }

    }
}


-(int) addText: (NSString*)data;
{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }
        int result = [epos2Printer_ addText: data];
        return result;
    }
}

-(int) addTextLang:(int)lang;
{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        int result = [epos2Printer_ addTextLang: lang];
        return result;
    }
}

-(int) addFeedLine: (int)line;
{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        int result = [epos2Printer_ addFeedLine: line];
        return result;
    }
}

-(int) addLineSpace: (int)linespc;
{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        int result = [epos2Printer_ addLineSpace: linespc];
        return result;
    }
}

-(int) addCommand: (NSString* )base64string;
{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        NSData *data = [[NSData alloc] initWithBase64EncodedString: base64string options:0];

        int result = [epos2Printer_ addCommand: data];
        return result;
    }
}

-(int) addPulse: (int)drawer time:(int)time;
{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        int result = [epos2Printer_ addPulse: drawer time:time];
        return result;
    }
}

-(int)addTextAlign:(int)align;

{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        int result = [epos2Printer_ addTextAlign: align];
        return result;
    }
}


-(int)addTextSize:(long)width height:(long)height;

{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        int result = [epos2Printer_ addTextSize: width height:height];
        return result;
    }
}

-(int)addTextSmooth:(int)smooth;

{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        int result = [epos2Printer_ addTextSmooth: smooth];
        return result;
    }
}

-(int) addTextStyle:(int)reverse ul:(int)ul em:(int)em color:(int)color;

{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        int result = [epos2Printer_ addTextStyle: reverse ul:ul em:em color:color];
        return result;
    }
}

-(int) addCut: (int)type;
{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        int result = [epos2Printer_ addCut: type];
        return result;
    }
}


-(int) addImage: (NSDictionary *)source
      width:(long)width
      color:(int)color
      mode:(int)mode
      halftone:(int)halftone
      brightness:(double)brightness
      compress:(int)compress;
{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }
        UIImage *data = [ImageManager getImageFromDictionarySource:source];
        CGSize size = [ImageManager getImageCGSize:data width:width];
        UIImage *scaledImage = [ImageManager scaleImage:data size:size];

        int result = [epos2Printer_ addImage: scaledImage x:0 y:0 width:size.width height:size.height color:color mode:mode halftone:halftone brightness:brightness compress:compress];
        return result;
    }
}

-(int) addBarcode: (NSString *)data type:(int)type hri:(int)hri font:(int)font width:(long)width height:(long)height;
{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        int result = [epos2Printer_ addBarcode: data type:type hri:hri font:font width:width height:height];
        return result;
    }
}

-(int) addSymbol:(NSString *)data type:(int)type level:(int)level width:(long)width height:(long)height size:(long)size;
{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        int result = [epos2Printer_ addSymbol: data type:type level:level width:width height:height size:size];
        return result;
    }
}

-(NSDictionary*) getStatus
{
    @synchronized (self) {
        Epos2PrinterStatusInfo* status = [epos2Printer_ getStatus];

        if(status) {
            NSDictionary* statusDict = [EposStringHelper convertStatusInfoToDictionary:status];
            return statusDict;
        } else {
            return nil;
        }
    }
}

-(void) getPrinterSetting: (long)timeout
        type:(int)type
        successHandler: (void(^)(NSDictionary* data)) successHandler
        errorHandler: (void(^)(NSString* data)) errorHandler
{
    @synchronized (self) {
        if (epos2Printer_ == nil) {
            errorHandler([EposStringHelper convertDictionatyToJsonString:@{
                @"data": @(EPOS2_ERR_MEMORY),
                @"type": @"result"
            }]);
            return;
        }

        [self beginTransaction];
        int result = [epos2Printer_ getPrinterSetting:timeout type:type delegate:self];

        if(result == EPOS2_SUCCESS) {
            onGetPrinterSettingSuccessHandler_ = [successHandler copy];
            onGetPrinterSettingErrorHandler_ = [errorHandler copy];
        } else {
            [self endTransaction];
            errorHandler(
             [EposStringHelper convertDictionatyToJsonString:@{
                @"data": @(result),
                @"type": @"result"
             }]);
        }
    }
}

-(int) beginTransaction;
{
    @synchronized (self) {

        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        if (!isConnected_) {
            return EPOS2_ERR_DISCONNECT;
        }

        int result = [epos2Printer_ beginTransaction];
        if (result != EPOS2_SUCCESS) {
            didBeginTransaction_ = false;
        } else {
            didBeginTransaction_ = true;
        }

        return result;

    }
}

-(int) endTransaction;
{
    @synchronized (self) {

        if (epos2Printer_ == nil) {
            return EPOS2_ERR_MEMORY;
        }

        if (!isConnected_) {
            return EPOS2_ERR_DISCONNECT;
        }

        if (!didBeginTransaction_) return EPOS2_SUCCESS;

        int result = [epos2Printer_ endTransaction];
        if (result != EPOS2_SUCCESS) {
            NSLog(@"Transaction end error!");
        } else {
            didBeginTransaction_ = false;
        }
        return result;

    }
}

#pragma mark - Epos2PtrReceiveDelegate
- (void) onPtrReceive:(Epos2Printer *)printerObj code:(int)code status:(Epos2PrinterStatusInfo *)status printJobId:(NSString *)printJobId
{
    @synchronized (self) {
        NSDictionary* returnData = [EposStringHelper convertStatusInfoToDictionary:status];
        [self endTransaction];

        if(onPtrRecieveSuccessHandler_ && onPtrRecieveErrorHandler_) {
            if(code == EPOS2_CODE_SUCCESS) {
                onPtrRecieveSuccessHandler_(returnData);
            } else {
                NSDictionary* errorDataDic = @{
                    @"data": @(code),
                    @"type": @"code"
                };
                onPtrRecieveErrorHandler_([EposStringHelper convertDictionatyToJsonString:errorDataDic]);
            }
        }

        onPtrRecieveSuccessHandler_ = nil;
        onPtrRecieveErrorHandler_ = nil;
    }
}


#pragma mark - Epos2PrinterSettingDelegate
- (void) onGetPrinterSetting:(int)code type:(int)type value:(int)value
{

    @synchronized (self) {
      NSDictionary* returnData = @{
            @"type": [NSNumber numberWithInt:type],
            @"value": [NSNumber numberWithInt:value],
          };
        [self endTransaction];

        if(onGetPrinterSettingSuccessHandler_ && onGetPrinterSettingErrorHandler_) {
            if(code == EPOS2_CODE_SUCCESS) {
                onGetPrinterSettingSuccessHandler_(returnData);
            } else {
                NSDictionary* errorDataDic = @{
                    @"data": @(code),
                    @"type": @"code"
                };
                onGetPrinterSettingErrorHandler_([EposStringHelper convertDictionatyToJsonString:errorDataDic]);
            }
        }

        onGetPrinterSettingSuccessHandler_ = nil;
        onGetPrinterSettingErrorHandler_ = nil;
    }
}

- (void) onSetPrinterSetting:(int)code
{

}
@end
